### PR TITLE
fixed inaccurate vehicle status change

### PR DESCRIPTION
**Pull request by pfreema1**
## Description

For a vehicles' status to change from `travelling_pickup` to `landing_pickup`, only the elapsed time is being taken into account to determine if the vehicle is ready to change to the next status (`landing_pickup`).  This could be the cause of this bug: 
 (https://github.com/DAVFoundation/missions/issues/145).

I fixed this issue by:
1. Adding the `vehicle` object as data to be passed into the `conditionForNextUpdate` function (StatusController.js).

2.  Using the vehicle coordinates in the `conditionForNextUpdate` function (missionProgress.js), to build an array of the last 3 coordinates and checking to see if they are the same.  If they are the same and the `mission.time_to_pickup` has elapsed, then the next vehicle status is set to `landing_pickup`.

If I can make this fix better let me know!  😄 

## Related Issue

link to the issue:   https://github.com/DAVFoundation/missions/issues/145

Steps to reproduce the bug:
1.  Order a drone, preferably one with > 1 minute pickup time so you can see the bug. 
2.  The Ready for Pickup UI, which is controlled by the vehicles' status changing from `travelling_pickup` to `landing_pickup`, appears too early.

## Motivation and Context

This will make the UI representation of the data more accurate.

## How Has This Been Tested?

-ran `npm test`
-tested with 5 different address combinations

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: server/controllers/StatusController.js [hunk 1]
```diff
@@ -42,11 +42,19 @@ const getStatus = async (req, res) => {
...
       const mission = latestMission;
       let vehicle = await getVehicle(latestMission.vehicle_id);
+      let runNextUpdate;
       const status = 'in_mission';
       const currentStatus = missionProgress[vehicle.status];
 
       if (currentStatus.beforeUpdate) await currentStatus.beforeUpdate(latestMission);
-      if (currentStatus.conditionForNextUpdate(latestMission)) {
+
+      if(currentStatus.status === 'travelling_pickup') {
```
Comment (srfrnk) [inline server/controllers/StatusController.js]: Why not send vehicle in both cases? Make the code look cleaner?

File: server/simulation/missionProgress.js [hunk 2]
```diff
@@ -1,13 +1,42 @@
 const { updateMission } = require('../store/missions');
 
+let prevCoords = [];
```
Comment (srfrnk) [inline server/simulation/missionProgress.js]: What would that hold when multiple vehicles are doing mission at the same time?

File: server/simulation/missionProgress.js [hunk 3]
```diff
@@ -1,13 +1,42 @@
...
+
+const prevCoordsSame = prevCoords => {
```
Comment (srfrnk) [inline server/simulation/missionProgress.js]: Why use prevCoords? Can't we just compare current location to target location?

File: server/simulation/missionProgress.js [hunk 4]
```diff
@@ -1,13 +1,42 @@
...
     status: 'travelling_pickup',
     nextVehicleStatus: 'landing_pickup',
     nextMissionUpdate: 'landing_pickup',
-    conditionForNextUpdate: mission => {
+    conditionForNextUpdate: (mission, vehicle) => {
       let elapsedTime = Date.now() - (parseFloat(mission.user_signed_at) + parseFloat(mission.time_to_pickup));
-      return elapsedTime > 0;
+      let isAtTerminal = false;
+
+      if(prevCoords.length < 3) {
```
Comment (srfrnk) [inline server/simulation/missionProgress.js]: I think maybe using current vs. target location would be better here - WDYT?

File: server/simulation/missionProgress.js [hunk 5]
```diff
@@ -5,9 +5,15 @@ module.exports = {
     status: 'travelling_pickup',
     nextVehicleStatus: 'landing_pickup',
     nextMissionUpdate: 'landing_pickup',
-    conditionForNextUpdate: mission => {
+    conditionForNextUpdate: (mission, vehicle) => {
       let elapsedTime = Date.now() - (parseFloat(mission.user_signed_at) + parseFloat(mission.time_to_pickup));
-      return elapsedTime > 0;
+      let isAtTerminal = false;
+
+      if(vehicle.coords.lat == mission.pickup_latitude && vehicle.coords.long == mission.pickup_longitude) {
```
Comment (srfrnk) [inline server/simulation/missionProgress.js]: That's great.
However float number comparison with === is most likely to fail (e.g. 0.00000001 !== 0.00000002).
Can you use the previous method of using a Diff Delta?

Comment (pfreema1) [reply server/simulation/missionProgress.js]: For sure! is `0.00000005` as the diff delta ok?

Comment (srfrnk) [reply server/simulation/missionProgress.js]: I think 10e-4 would be enough :)

Comment (srfrnk) [reply server/simulation/missionProgress.js]: And sorry for the delay - didn't notice your comment.

Comment (pfreema1) [reply server/simulation/missionProgress.js]: If 1x10e-4 is .0001, isn't that still a large distance when using GPS coords?  If I use a deltaThreshold of 10e-8 then it works correctly.  I'm getting the delta amount like so for long and lat:
`Math.abs(vehicleLat - pickupLat)` and `Math.abs(vehicleLong - pickupLong)`
Not sure if maybe there is a better way of finding the distance between two GPS coords.

Comment (srfrnk) [reply server/simulation/missionProgress.js]: Sounds good to me :)

**Pull request merged**